src/invidious/routes/API/v1/routes.cr

```diff
@@ -0,0 +1,30 @@
+# There is far too many API routes to define in invidious.cr
+# so we'll just do it here instead with a macro.
+macro define_v1_api_routes(base_url = "/api/v1")
+  Invidious::Routing.get "#{{{base_url}}}/stats", Invidious::Routes::V1Api, :stats
+
+  Invidious::Routing.get "#{{{base_url}}}/storyboards/:id", Invidious::Routes::V1Api, :storyboards
+  Invidious::Routing.get "#{{{base_url}}}/captions/:id", Invidious::Routes::V1Api, :captions
+  Invidious::Routing.get "#{{{base_url}}}/annotations/:id", Invidious::Routes::V1Api, :annotations
+  Invidious::Routing.get "#{{{base_url}}}/search/suggestions/:id", Invidious::Routes::V1Api, :search_suggestions
+
+  Invidious::Routing.get "#{{{base_url}}}/comments/:id", Invidious::Routes::V1Api, :comments
+  Invidious::Routing.get "#{{{base_url}}}/trending", Invidious::Routes::V1Api, :trending
+  Invidious::Routing.get "#{{{base_url}}}/popular", Invidious::Routes::V1Api, :popular
+
+  Invidious::Routing.get "#{{{base_url}}}/channels/:ucid", Invidious::Routes::V1Api, :home
+
+  {% for route in {
+                    {"home", "home"},
+                    {"videos", "videos"},
+                    {"latest", "latest"},
+                    {"playlists", "playlists"},
+                    {"comments", "community"}, # Why is the route for the community API `comments`?,
+                    {"search", "channel_search"},
+                  } %}
+
+  Invidious::Routing.get "#{{{base_url}}}/channels/#{{{route[0]}}}/:ucid", Invidious::Routes::V1Api, :{{route[1]}}
+  Invidious::Routing.get "#{{{base_url}}}/channels/:ucid/#{{{route[0]}}}", Invidious::Routes::V1Api, :{{route[1]}}
+
+  {% end %}
```

Comment:
instead of a macro in here, you could do something like that:

``` crystal
get "#{{{base_url}}}/channels/:route/:ucid", Invidious::Routes::V1Api, :channel_routes_handler
get "#{{{base_url}}}/channels/:ucid/:route", Invidious::Routes::V1Api, :channel_routes_handler
```

``` crystal
def channel_routes_handler(env)
  switch env.params.url["route"]
  case "home"
    home(env)
  case "videos"
    videos(env)
  # More cases
  else
    # 404 error
  end
end  
```

Reply:
Unfortunately due to an upstream bug with kemal this wouldn't work; any attempt to access `env.params.url[]` would just yield a 404. We'll have to use a workaround similar to this: https://github.com/iv-org/invidious/blob/3e5c353298df54463558d5f9f1ad0698fe145c74/src/invidious.cr#L1592 And change all of the channel API handlers to either implement the same or accept the ucid as a separate parameter.  

It really isn't that much work but I'd rather not use a feature that's currently broken upstream unless we absolutely have to.

Reply:
Ohhh, yeah, just realized. I'd say that it isn't a bug: kemal doesn't (and can't possibly) know what we expect to be an UCID and what we expect to be a route.

Then, the solution is to do the following:
```crystal
get "#{{{base_url}}}/channels/:ucid/:route", Invidious::Routes::V1Api, :channel_routes_handler
```
```crystal
def channel_routes_handler(env)
  # Handle API calls made by confused programmers who inverted
  # the channel route and the UCID.
  if env.params.url["route"].matches?(/UC[a-zA-Z0-9_-]{22})/)
    # Inverted
    route = env.params.url["ucid"]
    ucid = env.params.url["route"]
  elseif env.params.url["ucid"].matches?(/UC[a-zA-Z0-9_-]{22})/)
    # Normal
    route = env.params.url["route"]
    ucid = env.params.url["ucid"]
  else
    # Invalid UCID!
    raise # SomeException
  end

  switch route
    # Different routes handlers
  end
end
```

Reply:
> Ohhh, yeah, just realized. I'd say that it isn't a bug: kemal doesn't...

That's one part of it, but it's not the bug I'm referring to. I'm talking about the errror that occurs when routes like this occurs: `/a/:a`, `/a/:a/:b`, that prevents us from accessing anything in `env.params.url` without causing a 404 — as in http status event not missing hash key.

In our case, the route `.../channels/:ucid/:route` proceeds `.../channels/:ucid` and thus this strange behavior occurs.

Reply:
but `/api/v1/channels/:ucid`, doesn't exist, no?

Reply:
It does. https://github.com/iv-org/invidious/blob/c6a3ac9675f814baf49d9f272f10231110a6910e/src/invidious/routes/API/v1/routes.cr#L15